peers/app_request_network.go

```diff
@@ -126,95 +133,155 @@ func NewNetwork(
+				updateValidatorSet(context.Background(), n.logger, subnet, n.manager, n.validatorClient)
+			}
+			n.lock.Unlock()
+		}
+	}()
+}
 
-	startInfoAPICall := time.Now()
-	// Get the list of publicly discoverable peers
-	peers, err := n.infoAPI.Peers(context.Background(), nil)
-	n.setInfoAPICallLatencyMS(float64(time.Since(startInfoAPICall).Milliseconds()))
+func updateValidatorSet(
+	ctx context.Context,
+	logger logging.Logger,
+	subnetID ids.ID,
+	manager vdrs.Manager,
+	client *validators.CanonicalValidatorClient,
+) error {
+	logger.Debug("Fetching validators for subnet ID", zap.Stringer("subnetID", subnetID))
+
+	// Fetch the primary network validators from the P-Chain
+	height, err := client.GetCurrentHeight(ctx)
 	if err != nil {
-		n.logger.Error(
-			"Failed to get peers",
-			zap.Error(err),
-		)
-		return nil
+		return err
+	}
+	validators, err := client.GetValidatorSet(ctx, height, subnetID)
+	if err != nil {
+		return err
 	}
 
-	// Add manually tracked peers
-	peers = append(peers, n.manuallyTrackedPeers...)
+	validatorsMap := make(map[ids.NodeID]*vdrs.GetValidatorOutput)
+	for _, vdr := range validators {
+		validatorsMap[vdr.NodeID] = vdr
+	}
 
-	// Attempt to connect to each peer
-	var trackedNodes set.Set[ids.NodeID]
-	for _, peer := range peers {
-		if nodeIDs.Contains(peer.ID) {
-			trackedNodes.Add(peer.ID)
-			n.network.ManuallyTrack(peer.ID, peer.PublicIP)
-			if len(trackedNodes) == nodeIDs.Len() {
-				return trackedNodes
+	// Remove any elements from the manager that are not in the new validator set
+	currentVdrs := manager.GetValidatorIDs(subnetID)
+	for _, nodeID := range currentVdrs {
+		if _, ok := validatorsMap[nodeID]; !ok {
+			logger.Debug("Removing validator", zap.Stringer("nodeID", nodeID), zap.Stringer("subnetID", subnetID))
+			weight := manager.GetWeight(subnetID, nodeID)
+			if err := manager.RemoveWeight(subnetID, nodeID, weight); err != nil {
+				return err
 			}
 		}
 	}
 
-	// If the Info API node is in nodeIDs, it will not be reflected in the call to info.Peers.
-	// In this case, we need to manually track the API node.
-	startInfoAPICall = time.Now()
-	apiNodeID, _, err := n.infoAPI.GetNodeID(context.Background())
-	n.setInfoAPICallLatencyMS(float64(time.Since(startInfoAPICall).Milliseconds()))
-	if err != nil {
-		n.logger.Error(
-			"Failed to get API Node ID",
-			zap.Error(err),
-		)
-	} else if nodeIDs.Contains(apiNodeID) {
-		startInfoAPICall = time.Now()
-		apiNodeIPPort, err := n.infoAPI.GetNodeIP(context.Background())
-		n.setInfoAPICallLatencyMS(float64(time.Since(startInfoAPICall).Milliseconds()))
-		if err != nil {
-			n.logger.Error(
-				"Failed to get API Node IP",
-				zap.Error(err),
-			)
-		} else {
-			trackedNodes.Add(apiNodeID)
-			n.network.ManuallyTrack(apiNodeID, apiNodeIPPort)
+	// Add any elements from the new validator set that are not in the manager
+	for _, vdr := range validators {
+		if _, ok := manager.GetValidator(subnetID, vdr.NodeID); !ok {
```

Comment:
just confirming but I assume it's okay to never need to update the validators in the managers here manually?

Reply:
Not sure I follow. The result of this function call should be that the manager matches the validator set retreived from the P-Chain API.

Reply:
I might be missing something but I was wondering about the weight change cases. I see the removal and addition cases for validator mismatches between the P-Chain API and the manager but I also see that the manager keeps track of weights, and I don't see the case where they both exist but have different weights (if that's possible)

Weights can change through delegation while the validator is still active so I wasn't sure if there was a need to keep the manager state somehow manually in sync

Reply:
The validator manager is used purely for peer discovery, so the weights don't matter here. A validator's weight doesn't affect the way it's gossiped to peers AFAIK.